smol/moca/sampler/kernel.py

```diff
@@ -408,6 +407,272 @@ def single_step(self, occupancy):
+            exponent = entropy - new_entropy
+            self.trace.accepted = np.array(True if exponent >= 0 else
+                                           exponent > log(self._rng.random()))
+
+        if self.trace.accepted:
+            for f in step:
+                occupancy[f[0]] = f[1]
+            bin_id = new_bin_id
+            self._current_features += self.trace.delta_trace.features
+            self._current_enthalpy = new_enthalpy
+            self._usher.update_aux_state(step)
+        self.trace.occupancy = occupancy
+
+        # Only if bin_id is valid, because bin_id is not checked.
+        if 0 <= bin_id < len(self._levels):
+            # compute the cumulative statistics
+            self._states_counter += 1
+            total = self._occurrences[bin_id]
+            self._mean_features[bin_id, :] = \
+                1 / (total + 1) * (self._current_features
+                                   + total * self._mean_features[bin_id, :]
+                                   )
+            # check if histograms are flat and reset accordingly
+            if self._states_counter % self.update_period == 0:
+                # Add to DOS and histogram
+                self._entropy[bin_id] += self._m
+                self._histogram[bin_id] += 1
+                self._occurrences[bin_id] += 1
+
+        # fill up values in the trace
+        self.trace.histogram = self._histogram
+        self.trace.occurrences = self._occurrences
+        self.trace.entropy = self._entropy
+        self.trace.mean_features = self._mean_features
+        self.trace.mod_factor = np.array([self._m])
+
+        self._steps_counter += 1
+        if self._steps_counter % self.check_period == 0:
+            histogram = self._histogram[self._histogram > 0]  # remove zero entries
+            if ((histogram > self.flatness * histogram.mean()).all()
+                    and len(histogram) >= 0.5 * len(self._histogram)):
+                # A sufficient portion of histogram must have been populated.
+                # So be careful with your min and max energy setting.
+                self._m = self._mod_update(self._m)
+                self._histogram[:] = 0  # reset histogram and decrease _m.
+
+        return self.trace
+
+    def compute_initial_trace(self, occupancy):
+        """Compute initial values for sample trace given an occupancy.
+
+        Add the micro-canonical entropy and histograms to the trace.
+        Args:
+            occupancy (ndarray):
+                Initial occupancy
+        Returns:
+            Trace
+        """
+        trace = super().compute_initial_trace(occupancy)
+        # This will not be counting the state corresponding to the given occupancy.
+        # but that is ok since we are not recording the initial trace when sampling.
+        trace.histogram = self._histogram
+        trace.occurrences = self._occurrences
+        trace.entropy = self._entropy
+        trace.mean_features = self._mean_features
```

Comment:
can we call it `trace.cumulative_mean_features` so that the user interface is more explicit on what this is?

Reply:
The private variable `_mean_features` is ok since that's not part of the public interface.